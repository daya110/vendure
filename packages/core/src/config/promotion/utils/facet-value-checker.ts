import { ID } from '@vendure/common/lib/shared-types';
import { unique } from '@vendure/common/lib/unique';
import { Connection } from 'typeorm';

import { TtlCache } from '../../../common/ttl-cache';
import { idsAreEqual } from '../../../common/utils';
import { OrderLine } from '../../../entity/order-line/order-line.entity';
import { ProductVariant } from '../../../entity/product-variant/product-variant.entity';

export class FacetValueChecker {
    private variantCache = new TtlCache<ID, ProductVariant>({ ttl: 5000 });

    constructor(private connection: Connection) {}
    /**
     * @description
     * Checks a given {@link OrderLine} against the facetValueIds and returns
     * `true` if the associated {@link ProductVariant} & {@link Product} together
     * have *all* the specified {@link FacetValue}s.
     */
    async hasFacetValues(orderLine: OrderLine, facetValueIds: ID[]): Promise<boolean> {
        let variant = this.variantCache.get(orderLine.productVariant.id);
        if (!variant) {
            variant = await this.connection
                .getRepository(ProductVariant)
                .findOne(orderLine.productVariant.id, {
                    relations: ['product', 'product.facetValues', 'facetValues'],
                });
            if (!variant) {
                return false;
            }
            this.variantCache.set(variant.id, variant);
        }
        const allFacetValues = unique([...variant.facetValues, ...variant.product.facetValues], 'id');
        return facetValueIds.reduce(
            (result, id) => result && !!allFacetValues.find(fv => idsAreEqual(fv.id, id)),
            true as boolean,
        );
    }
}
