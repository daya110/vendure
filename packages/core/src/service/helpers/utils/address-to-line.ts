import { OrderAddress } from '@vendure/common/lib/generated-types';

import { Address } from '../../../entity/address/address.entity';

/**
 * Given an Address object, this function converts it into a single line
 * consisting of streetLine1, (postalCode), (countryCode).
 */
export function addressToLine(address: Address | OrderAddress): string {
    const propsToInclude: Array<keyof (Address | OrderAddress)> = ['streetLine1', 'postalCode', 'country'];
    let result = address.streetLine1 || '';
    if (address.postalCode) {
        result += ', ' + address.postalCode;
    }
    if (address.country) {
        if (typeof address.country === 'string') {
            result += ', ' + address.country;
        } else {
            result += ', ' + address.country.name;
        }
    }
    return result;
}
