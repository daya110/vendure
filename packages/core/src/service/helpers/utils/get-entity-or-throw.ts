import { ID, Type } from '@vendure/common/shared-types';
import { Connection, FindOneOptions } from 'typeorm';

import { EntityNotFoundError } from '../../../common/error/errors';
import { SoftDeletable } from '../../../common/types/common-types';
import { VendureEntity } from '../../../entity/base/base.entity';

/**
 * Attempts to find an entity of the given type and id, and throws an error if not found.
 */
export async function getEntityOrThrow<T extends VendureEntity>(
    connection: Connection,
    entityType: Type<T>,
    id: ID,
    findOptions?: FindOneOptions<T>,
): Promise<T> {
    const entity = await connection.getRepository(entityType).findOne(id, findOptions);
    if (!entity || (entity.hasOwnProperty('deletedAt') && (entity as T & SoftDeletable).deletedAt !== null)) {
        throw new EntityNotFoundError(entityType.name as any, id);
    }
    return entity;
}
