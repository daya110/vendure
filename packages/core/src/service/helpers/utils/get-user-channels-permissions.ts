import { Permission } from '@vendure/common/lib/generated-types';
import { ID } from '@vendure/common/lib/shared-types';
import { unique } from '@vendure/common/lib/unique';

import { User } from '../../../entity/user/user.entity';

export interface UserChannelPermissions {
    id: ID;
    token: string;
    code: string;
    permissions: Permission[];
}

/**
 * Returns an array of Channels and permissions on those Channels for the given User.
 */
export function getUserChannelsPermissions(user: User): UserChannelPermissions[] {
    const channelsMap: { [code: string]: UserChannelPermissions } = {};

    for (const role of user.roles) {
        for (const channel of role.channels) {
            if (!channelsMap[channel.code]) {
                channelsMap[channel.code] = {
                    id: channel.id,
                    token: channel.token,
                    code: channel.code,
                    permissions: [],
                };
            }
            channelsMap[channel.code].permissions = unique([
                ...channelsMap[channel.code].permissions,
                ...role.permissions,
            ]);
        }
    }

    return Object.values(channelsMap).sort((a, b) => (a.id < b.id ? -1 : 1));
}
