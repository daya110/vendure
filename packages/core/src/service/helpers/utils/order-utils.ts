import { OrderItem } from '../../../entity/order-item/order-item.entity';
import { Order } from '../../../entity/order/order.entity';
import { PaymentState } from '../payment-state-machine/payment-state';

/**
 * Returns true if the Order total is covered by Payments in the specified state.
 */
export function orderTotalIsCovered(order: Order, state: PaymentState): boolean {
    return (
        order.payments.filter((p) => p.state === state).reduce((sum, p) => sum + p.amount, 0) === order.total
    );
}

/**
 * Returns true if all (non-cancelled) OrderItems are fulfilled.
 */
export function orderItemsAreFulfilled(order: Order) {
    return getOrderItems(order)
        .filter((orderItem) => !orderItem.cancelled)
        .every(isFulfilled);
}

/**
 * Returns true if at least one, but not all (non-cancelled) OrderItems are fulfilled.
 */
export function orderItemsArePartiallyFulfilled(order: Order) {
    const nonCancelledItems = getOrderItems(order).filter((orderItem) => !orderItem.cancelled);
    return nonCancelledItems.some(isFulfilled) && !nonCancelledItems.every(isFulfilled);
}

/**
 * Returns true if all OrderItems in the order are cancelled
 */
export function orderItemsAreAllCancelled(order: Order) {
    return getOrderItems(order).every((orderItem) => orderItem.cancelled);
}

function getOrderItems(order: Order): OrderItem[] {
    return order.lines.reduce((orderItems, line) => [...orderItems, ...line.items], [] as OrderItem[]);
}

function isFulfilled(orderItem: OrderItem) {
    return !!orderItem.fulfillment;
}
