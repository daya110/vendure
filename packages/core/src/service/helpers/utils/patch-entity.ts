import { VendureEntity } from '../../../entity/base/base.entity';

export type InputPatch<T> = { [K in keyof T]?: T[K] | null };

/**
 * Updates only the specified properties from an Input object as long as the value is not
 * undefined. Null values can be passed, however, which will set the corresponding entity
 * field to "null". So care must be taken that this is only done on nullable fields.
 */
export function patchEntity<T extends VendureEntity, I extends InputPatch<T>>(entity: T, input: I): T {
    for (const key of Object.keys(entity)) {
        const value = input[key as keyof T];
        if (key === 'customFields' && value) {
            patchEntity((entity as any)[key], value as any);
        } else if (value !== undefined && key !== 'id') {
            entity[key as keyof T] = value as any;
        }
    }
    return entity;
}
