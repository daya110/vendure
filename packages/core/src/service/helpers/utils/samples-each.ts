/**
 * Returns true if and only if exactly one item from each
 * of the "groups" arrays appears in the "sample" array.
 */
export function samplesEach<T>(sample: T[], groups: T[][]): boolean {
    if (sample.length !== groups.length) {
        return false;
    }
    return groups.every(group => {
        for (const item of sample) {
            if (group.includes(item)) {
                return true;
            }
        }
        return false;
    });
}
