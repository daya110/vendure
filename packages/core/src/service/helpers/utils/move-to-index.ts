import { Orderable } from '../../../common/types/common-types';
import { idsAreEqual } from '../../../common/utils';
import { VendureEntity } from '../../../entity/base/base.entity';

/**
 * Moves the target Orderable entity to the given index amongst its siblings.
 * Returns the siblings (including the target) which should then be
 * persisted to the database.
 */
export function moveToIndex<T extends Orderable & VendureEntity>(
    index: number,
    target: T,
    siblings: T[],
): T[] {
    const normalizedIndex = Math.max(Math.min(index, siblings.length), 0);
    let currentIndex = siblings.findIndex(sibling => idsAreEqual(sibling.id, target.id));
    const orderedSiblings = [...siblings].sort((a, b) => (a.position > b.position ? 1 : -1));
    const siblingsWithTarget = currentIndex < 0 ? [...orderedSiblings, target] : [...orderedSiblings];
    currentIndex = siblingsWithTarget.findIndex(sibling => idsAreEqual(sibling.id, target.id));
    if (currentIndex !== normalizedIndex) {
        siblingsWithTarget.splice(normalizedIndex, 0, siblingsWithTarget.splice(currentIndex, 1)[0]);
        siblingsWithTarget.forEach((collection, i) => {
            collection.position = i;
            if (target.id === collection.id) {
                target.position = i;
            }
        });
    }
    return siblingsWithTarget;
}
