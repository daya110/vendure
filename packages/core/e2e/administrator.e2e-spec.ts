import { createTestEnvironment } from '@vendure/testing';
import gql from 'graphql-tag';
import path from 'path';

import { initialData } from '../../../e2e-common/e2e-initial-data';
import { TEST_SETUP_TIMEOUT_MS, testConfig } from '../../../e2e-common/test-config';

import { ADMINISTRATOR_FRAGMENT } from './graphql/fragments';
import {
    Administrator,
    CreateAdministrator,
    GetAdministrator,
    GetAdministrators,
    UpdateAdministrator,
} from './graphql/generated-e2e-admin-types';
import { CREATE_ADMINISTRATOR } from './graphql/shared-definitions';
import { assertThrowsWithMessage } from './utils/assert-throws-with-message';

describe('Administrator resolver', () => {
    const { server, adminClient } = createTestEnvironment(testConfig);
    let createdAdmin: Administrator.Fragment;

    beforeAll(async () => {
        await server.init({
            initialData,
            productsCsvPath: path.join(__dirname, 'fixtures/e2e-products-minimal.csv'),
            customerCount: 1,
        });
        await adminClient.asSuperAdmin();
    }, TEST_SETUP_TIMEOUT_MS);

    afterAll(async () => {
        await server.destroy();
    });

    it('administrators', async () => {
        const result = await adminClient.query<GetAdministrators.Query, GetAdministrators.Variables>(
            GET_ADMINISTRATORS,
        );
        expect(result.administrators.items.length).toBe(1);
        expect(result.administrators.totalItems).toBe(1);
    });

    it('createAdministrator', async () => {
        const result = await adminClient.query<CreateAdministrator.Mutation, CreateAdministrator.Variables>(
            CREATE_ADMINISTRATOR,
            {
                input: {
                    emailAddress: 'test@test.com',
                    firstName: 'First',
                    lastName: 'Last',
                    password: 'password',
                    roleIds: ['1'],
                },
            },
        );

        createdAdmin = result.createAdministrator;
        expect(createdAdmin).toMatchSnapshot();
    });

    it('administrator', async () => {
        const result = await adminClient.query<GetAdministrator.Query, GetAdministrator.Variables>(
            GET_ADMINISTRATOR,
            {
                id: createdAdmin.id,
            },
        );
        expect(result.administrator).toEqual(createdAdmin);
    });

    it('updateAdministrator', async () => {
        const result = await adminClient.query<UpdateAdministrator.Mutation, UpdateAdministrator.Variables>(
            UPDATE_ADMINISTRATOR,
            {
                input: {
                    id: createdAdmin.id,
                    emailAddress: 'new-email',
                    firstName: 'new first',
                    lastName: 'new last',
                    password: 'new password',
                    roleIds: ['2'],
                },
            },
        );
        expect(result.updateAdministrator).toMatchSnapshot();
    });

    it('updateAdministrator works with partial input', async () => {
        const result = await adminClient.query<UpdateAdministrator.Mutation, UpdateAdministrator.Variables>(
            UPDATE_ADMINISTRATOR,
            {
                input: {
                    id: createdAdmin.id,
                    emailAddress: 'newest-email',
                },
            },
        );
        expect(result.updateAdministrator.emailAddress).toBe('newest-email');
        expect(result.updateAdministrator.firstName).toBe('new first');
        expect(result.updateAdministrator.lastName).toBe('new last');
    });

    it(
        'updateAdministrator throws with invalid roleId',
        assertThrowsWithMessage(
            () =>
                adminClient.query<UpdateAdministrator.Mutation, UpdateAdministrator.Variables>(
                    UPDATE_ADMINISTRATOR,
                    {
                        input: {
                            id: createdAdmin.id,
                            emailAddress: 'new-email',
                            firstName: 'new first',
                            lastName: 'new last',
                            password: 'new password',
                            roleIds: ['999'],
                        },
                    },
                ),
            `No Role with the id '999' could be found`,
        ),
    );
});

export const GET_ADMINISTRATORS = gql`
    query GetAdministrators($options: AdministratorListOptions) {
        administrators(options: $options) {
            items {
                ...Administrator
            }
            totalItems
        }
    }
    ${ADMINISTRATOR_FRAGMENT}
`;

export const GET_ADMINISTRATOR = gql`
    query GetAdministrator($id: ID!) {
        administrator(id: $id) {
            ...Administrator
        }
    }
    ${ADMINISTRATOR_FRAGMENT}
`;

export const UPDATE_ADMINISTRATOR = gql`
    mutation UpdateAdministrator($input: UpdateAdministratorInput!) {
        updateAdministrator(input: $input) {
            ...Administrator
        }
    }
    ${ADMINISTRATOR_FRAGMENT}
`;
