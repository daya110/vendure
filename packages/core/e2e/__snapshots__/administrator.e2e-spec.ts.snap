// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Administrator resolver createAdministrator 1`] = `
Object {
  "emailAddress": "test@test.com",
  "firstName": "First",
  "id": "T_2",
  "lastName": "Last",
  "user": Object {
    "id": "T_3",
    "identifier": "test@test.com",
    "lastLogin": null,
    "roles": Array [
      Object {
        "code": "__super_admin_role__",
        "description": "SuperAdmin",
        "id": "T_1",
        "permissions": Array [
          "Authenticated",
          "SuperAdmin",
          "CreateCatalog",
          "ReadCatalog",
          "UpdateCatalog",
          "DeleteCatalog",
          "CreateCustomer",
          "ReadCustomer",
          "UpdateCustomer",
          "DeleteCustomer",
          "CreateAdministrator",
          "ReadAdministrator",
          "UpdateAdministrator",
          "DeleteAdministrator",
          "CreateOrder",
          "ReadOrder",
          "UpdateOrder",
          "DeleteOrder",
          "CreatePromotion",
          "ReadPromotion",
          "UpdatePromotion",
          "DeletePromotion",
          "CreateSettings",
          "ReadSettings",
          "UpdateSettings",
          "DeleteSettings",
        ],
      },
    ],
  },
}
`;

exports[`Administrator resolver updateAdministrator 1`] = `
Object {
  "emailAddress": "new-email",
  "firstName": "new first",
  "id": "T_2",
  "lastName": "new last",
  "user": Object {
    "id": "T_3",
    "identifier": "new-email",
    "lastLogin": null,
    "roles": Array [
      Object {
        "code": "__customer_role__",
        "description": "Customer",
        "id": "T_2",
        "permissions": Array [
          "Authenticated",
        ],
      },
    ],
  },
}
`;
